text/0000-fn-pin-put.md

```diff
@@ -0,0 +1,189 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We add a new trait, `FnPinMut`, with the signature above, to `std::ops`.
+It is subject to the same restrictions as the stable `Fn` traits:
+it cannot be named directly (use the `FnPinMut()` form),
+and it cannot be implemented by hand.
+(These restrictions may be lifted in the future.)
+
+An implementation of `FnPinMut` is provided for all closures where said implementation is sound.
+Practically, this means all currently stable `FnMut` closures, as they are `Unpin`, get the following implementation:
+
+```rust
+impl FnPinMut<Args> for $GeneratedClosureType {
+    extern "rust-call" fn call_pin(self: Pin<&mut Self>, args: Args) -> Self::Output {
+        (&mut *self).call_mut(args)
+    }
+}
+```
+
+Closures are now allowed to contain **yield expression**s of the form `yield $expr`.
+If a closure contains a yield expression, it _may not_ also contain a return expression.
+This includes the trailing return: a closure containing a `yield` must "return" `!`.
+This will most often be accomplished by a `panic!` or a `loop` at the tail end.
+In most ways (allowed positions, return type), a yield statement acts the same as a return expression.
+
+We call a closure that uses `yield` instead of `return` a **yield closure**.
+
+The key difference is that rather than execution restarting at the top of
+the closure on the next call, execution resumes after the yield statement.
+The compiler will generate a state machine to jump to the correct place upon each resumation.
+
+This is subject to the normal rules about what `Fn` traits are implemented!
+This means that a yield closure can implment all four traits: `Fn`, `FnMut`, `FnPinMut`, and `FnOnce`!
+For example, [`std::iter`'s `Counter` example](counter) can be reimplemented as simply
+
+```rust
+std::iter::from_fn(|| {
+    for count in 1..6 {
+        yield Some(count);
+    }
+    loop {
+        yield None;
+    }
+})
+```
+
+This closure will implement all four `Fn` traits.
+
+So far, however, everything we've seen has been within the domain of the stable `Fn` traits.
+`FnPinMut` is required to support borrowing over `yield` points.
+
+(TODO)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+# Prior art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and the bad, in relation to this proposal.
+A few examples of what this can include are:
+
+- For language, library, cargo, tools, and compiler proposals: Does this feature exist in other programming languages and what experience have their community had?
```

Comment:
On my Reddit post, someone brought up `yield from` in python which allows one coroutine to delegate to another. Python's coroutines can signal that they have reached an end state by throwing `StopIteration`. When that happens, any delegation ends and `yield from` evaluates to the last `yield`ed value. That actually means that `return x` inside a python coroutine is the same as `yield x; throw StopIteration` which looks a lot like our potential desugar for `return`! The difference is that Python can then end the coroutine delegation while we would have no obvious way to do so (having gotten rid of `GeneratorState`). Just an interesting validation + trade-off.